Add doc about JsonSerializer collection support in 3.1 (#1886)
dotnet/runtime: src/libraries/System.Text.Json/docs/collection_support.md

# JsonSerializer collection support in 3.1.x

> All collections (any type that derives from IEnumerable) are supported for

Reviewer: Consider making all references to types/collections, links to the API docs for those types.
```suggestion
All collections (any type that derives from `IEnumerable`) are supported for
```

> serialization with JsonSerializer; we simply call the .GetEnumerator() method

Reviewer: ```suggestion
serialization with `JsonSerializer`; we simply call the `.GetEnumerator()` method
```

Deserialization is more complicated and is not supported for all collections.

This document gives a quick overview of which collections are supported.

## Built-in types

### System.Collections

| Type | Serialization | Deserialization |
| --- | --- | --- |
| ArrayList | Supported | Supported |
| BitArray | Supported | Not supported |
| Hashtable | Supported | Supported |
| Queue | Supported | Supported |
| SortedList | Supported | Supported |
| Stack | Supported | Supported |
| DictionaryEntry | Supported | Supported |
| ICollection | Supported | Supported |
| IDictionary | Supported | Supported |
| IEnumerable | Supported | Supported |
| IList | Supported | Supported |

### System.Collections.Generic

| Type | Serialization | Deserialization |
| --- | --- | --- |
| Dictionary<string, TValue> | Supported | Supported |
| HashSet<T> | Supported | Supported |
| LinkedList<T> | Supported | Supported |
| LinkedListNode<T> | Supported | Not supported |
| List<T> | Supported | Supported |
| Queue<T> | Supported | Supported |
| SortedDictionary<string,TValue> | Supported | Supported |
| SortedList<string,TValue> | Supported | Supported |
| SortedSet<T> | Supported | Supported |

> | Stack<T> | Supported | Supported |

Reviewer: Should we put an asterisk next to those where the behavior might be unexpected (so we can add notes about them below the table)? For example, serializing `Stack<T>` types.

| HybridDictionary | Supported | Supported |
| IOrderedDictionary | Supported | Not supported |
| ListDictionary | Supported | Supported |
| StringCollection | Supported | Not supported |
| StringDictionary | Supported | Not supported |
| NameValueCollection | Supported | Not supported |

### System.Collections.Concurrent

| Type | Serialization | Deserialization |
| --- | --- | --- |
| BlockingCollection<T> | Supported | Not supported |
| ConcurrentBag<T> | Supported | NotSupported |
| ConcurrentDictionary<string, TValue> | Supported | Supported |
| ConcurrentQueue<T> | Supported | Supported |
| ConcurrentStack<T> | Supported | Supported |

### System.Collections.ObjectModel

| Type | Serialization | Deserialization |
| --- | --- | --- |
| Collection<T> | Supported | Supported |
| ObservableCollection<T> | Supported | Supported |
| KeyedCollection<string, TValue> | Supported | Not supported |
| ReadOnlyCollection<T> | Supported | Not supported |
| ReadOnlyObservableCollection<T> | Supported | Not supported |
| ReadOnlyDictionary<string, TValue> | Supported | Not supported |

## Custom collections

For the purpose of serialization and deserialization in System.Text.Json, any

> collection not in the BCL, (i.e. outside System.Collections[.[.*]]), is considered

Reviewer: Why is everything after the regex italicized? Would back-ticks fix it?

```suggestion
collection not in the BCL, (i.e. outside `System.Collections[.[.*]]`), is considered
```

a custom collection. This includes user-defined types and ASP.NET defined types,
e.g. those in
[Microsoft.Extensions.Primitives](https://docs.microsoft.com/dotnet/api/microsoft.extensions.primitives?view=dotnet-plat-ext-3.1).

A custom collection is supported for deserialization if it fulfils the following:

1. Is not an interface or abstract.
2. Has a parameter-less constructor

> 3. Implements one or more of IList, IList<T>, ICollection<T>, Stack<T>, Queue<T>,

Reviewer: ```suggestion
3. Implements one or more of `IList`, `IList<T>`, `ICollection<T>`, `Stack<T>`, `Queue<T>`,
```

>    Stack, Queue, IDictionary, IDictionary<string, TValue>

Reviewer: ```suggestion
   `Stack`, `Queue`, `IDictionary`, `IDictionary<string, TValue>`
```

> 4. The element type is supported in System.Text.Json

Reviewer: Here and elsewhere.
```suggestion
4. The element type is supported in `System.Text.Json`
```

> All custom collections are supported for serialization (everything that derives from IEnumerable),

Reviewer: ```suggestion
All custom collections are supported for serialization (everything that derives from `IEnumerable`),
```

as long as their element types are supported.

> There are known issues with some custom collections where we don't offer rountrippable support.

Reviewer: ```suggestion
There are known issues with some custom collections where we don't offer round-trippable support.
```

These include:

- ExpandoObject: https://github.com/dotnet/corefx/issues/38007
- DynamicObject: https://github.com/dotnet/corefx/issues/41105 
- DataTable: https://github.com/dotnet/corefx/issues/38712 

> - Microsoft.AspNetCore.Http.FormFile: https://github.com/dotnet/corefx/issues/41401 

Reviewer: What about others:
https://github.com/dotnet/corefx/issues/42084
https://github.com/dotnet/corefx/issues/41105
Maybe? https://github.com/dotnet/corefx/issues/42399
Maybe? https://github.com/dotnet/corefx/issues/41399

| ICollection<T> | Supported | Supported |
| IDictionary<string,TValue> | Supported | Supported |
| IEnumerable<T> | Supported | Supported |
| IList<T> | Supported | Supported |
| IReadOnlyCollection<T> | Supported | Supported |
| IReadOnlyDictionary<string,TValue> | Supported | Supported |
| IReadOnlyList<T> | Supported | Supported |
| ISet<T> | Supported | Supported |

### System.Collections.Immutable

| Type | Serialization | Deserialization |
| --- | --- | --- |
| ImmutableArray<T> | Supported | Supported |
| ImmutableDictionary<string, TValue> | Supported | Supported |
| ImmutableHashSet<T> | Supported | Supported |
| ImmutableList<T> | Supported | Supported |
| ImmutableQueue<T> | Supported | Supported |
| ImmutableSortedDictionary<string, TValue> | Supported | Supported |
| ImmutableSortedSet<T> | Supported | Supported |
| ImmutableStack<T> | Supported | Supported |
| IImmutableDictionary<string, TValue> | Supported | Supported |
| IImmutableList<T> | Supported | Supported |
| IImmutableQueue<T> | Supported | Supported |
| IImmutableSet<T> | Supported | Supported |
| IImmutableStack<T> | Supported | Supported |

### System.Collections.Specialized

> #### Deserialization heuristics

Reviewer 2: What does this mean?

Author: I had notes earlier about how the deserialization is done for each class of types. This is not relevant is this doc so I removed them (and forgot this one). I just pushed a commit to remove this line.

| Type | Serialization | Deserialization |
| --- | --- | --- |
| [`ImmutableArray<T>`](https://docs.microsoft.com/dotnet/api/system.collections.immutable.immutablearray-1?view=netcore-3.1) | Supported | Supported |
| [`ImmutableDictionary<string, TValue>`](https://docs.microsoft.com/dotnet/api/system.collections.immutable.immutabledictionary-2?view=netcore-3.1) | Supported | Supported |
| [`ImmutableHashSet<T>`](https://docs.microsoft.com/dotnet/api/system.collections.immutable.immutablehashset-1?view=netcore-3.1) | Supported | Supported |
| [`IImmutableList<T>`](https://docs.microsoft.com/dotnet/api/system.collections.immutable.iimmutablelist-1?view=netcore-3.1) | Supported | Supported |
| [`ImmutableQueue<T>`](https://docs.microsoft.com/dotnet/api/system.collections.immutable.immutablequeue-1?view=netcore-3.1) | Supported | Supported |
| [`ImmutableSortedDictionary<string, TValue>`](https://docs.microsoft.com/dotnet/api/system.collections.immutable.immutablesorteddictionary-2?view=netcore-3.1) | Supported | Supported |
| [`ImmutableSortedSet<T>`](https://docs.microsoft.com/dotnet/api/system.collections.immutable.immutablesortedset-1?view=netcore-3.1) | Supported | Supported |

> | [`ImmutableStack<T>`](https://docs.microsoft.com/dotnet/api/system.collections.immutable.immutablestack-1?view=netcore-3.1) | Supported | Supported |

Reviewer: Add `*` to these other Stack types

These include:

- Support for [`ExpandoObject`](https://docs.microsoft.com/dotnet/api/system.dynamic.expandoobject?view=netcore-3.1): https://github.com/dotnet/corefx/issues/38007
- Support for [`DynamicObject`](https://docs.microsoft.com/dotnet/api/system.dynamic.dynamicobject?view=netcore-3.1): https://github.com/dotnet/corefx/issues/41105
- Support for [`DataTable`](https://docs.microsoft.com/dotnet/api/system.data.datatable?view=netcore-3.1): https://github.com/dotnet/corefx/issues/38712
- Support for [`FormFile`](https://docs.microsoft.com/dotnet/api/microsoft.aspnetcore.http.formfile?view=aspnetcore-3.1): https://github.com/dotnet/corefx/issues/41401
- Support for [`IFormCollection`](https://docs.microsoft.com/dotnet/api/microsoft.aspnetcore.http.iformcollection?view=aspnetcore-3.1)
- Assigning `null` to value-type collections like [`ImmutableArray<T>`](https://docs.microsoft.com/dotnet/api/system.collections.immutable.immutablearray-1?view=netcore-3.1): https://github.com/dotnet/corefx/issues/42399

For more information, see the [open issues in System.Text.Json](https://github.com/dotnet/runtime/issues?q=is%3Aopen+is%3Aissue+label%3Aarea-System.Text.Json).

> \* All instances of custom collections that implement stack types are reversed on serialization,

Reviewer: Was repeating this intentional?

Author: Yes, there was a distinction between built-in types and custom types. I've consolidated these notes at the bottom.
